### PR TITLE
feat: emit Kafka stats as metrics instead of logs

### LINKED ISSUE #3100: Log is being spammed with rdkafka `Client stats` messages
# Rationale
Logging useless messages consumes resources (and time to filter them out while looking for needles)

I was looking at IOx logs on greylog and I found it was generating 140,000 such messages per hour (~39 per *second*!!), which looks like approximately 30% - 50% of all messages from IOx.

Unfortunately, it appears that the logs are coming from the `rdkafka` module: /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/client.rs log.line=85 target="log:

Screen Shot 2021-11-13 at 7 08 17 AM

Example log message:
```
level=info msg="Client stats: Statistics { name: \"rdkafka#consumer-11\", client_id: \"rdkafka\", client_type: \"consumer\", ts: 686505726551, time: 1636805420, replyq: 0, msg_cnt: 0, msg_size: 0, msg_max: 0, msg_size_max: 0, tx: 840044, tx_bytes: 103095761, rx: 840043, rx_bytes: 87137006, txmsgs: 0, txmsg_bytes: 0, rxmsgs: 0, rxmsg_bytes: 0, simple_cnt: 0, metadata_cache_cnt: 1, brokers: {\"iox-kafka-1.iox-kafka-brokers.iox.svc:9092/1\": Broker { name: \"iox-kafka-1.iox-kafka-brokers.iox.svc:9092/1\", nodeid: 1, nodename: \"iox-kafka-1.iox-kafka-brokers.iox.svc:9092\", source: \"learned\", state: \"INIT\", stateage: 418927316889, outbuf_cnt: 0, outbuf_msg_cnt: 0, waitresp_cnt: 0, waitresp_msg_cnt: 0, tx: 0, txbytes: 0, txerrs: 0, txretries: 0, req_timeouts: 0, rx: 0, rxbytes: 0, rxerrs: 0, rxcorriderrs: 0, rxpartial: 0, req: {\"JoinGroup\": 0, \"OffsetFetch\": 0, \"Heartbeat\": 0, \"DescribeClientQuotasRequest\": 0, \"Offset\": 0, \"FindCoordinator\": 0, \"OffsetDeleteRequest\": 0, \"ApiVersion\": 0, \"Fetch\": 0, \"AlterClientQuotasRequest\": 0, \"SaslAuthenticate\": 0, \"SaslHandshake\": 0, \"SyncGroup\": 0, \"DescribeUserScramCredentialsRequest\": 0, \"Metadata\": 0, \"LeaveGroup\": 0, \"OffsetCommit\": 0}, zbuf_grow: 0, buf_grow: 0, wakeups: Some(0), connects: Some(0), disconnects: Some(0), int_latency: Some(Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 11376, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }), outbuf_latency: Some(Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 11376, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }), rtt: Some(Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 13424, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }), throttle: Some(Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 17520, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }), toppars: {} }, \"iox-kafka-0.iox-kafka-brokers.iox.svc:9092/0\": Broker { name: \"iox-kafka-0.iox-kafka-brokers.iox.svc:9092/0\", nodeid: 0, nodename: \"iox-kafka-0.iox-kafka-brokers.iox.svc:9092\", source: \"learned\", state: \"INIT\", stateage: 418927317492, outbuf_cnt: 0, outbuf_msg_cnt: 0, waitresp_cnt: 0, waitresp_msg_cnt: 0, tx: 0, txbytes: 0, txerrs: 0, txretries: 0, req_timeouts: 0, rx: 0, rxbytes: 0, rxerrs: 0, rxcorriderrs: 0, rxpartial: 0, req: {\"Fetch\": 0, \"DescribeUserScramCredentialsRequest\": 0, \"SaslHandshake\": 0, \"ApiVersion\": 0, \"Offset\": 0, \"SaslAuthenticate\": 0, \"OffsetCommit\": 0, \"LeaveGroup\": 0, \"OffsetDeleteRequest\": 0, \"DescribeClientQuotasRequest\": 0, \"Metadata\": 0, \"OffsetFetch\": 0, \"SyncGroup\": 0, \"FindCoordinator\": 0, \"JoinGroup\": 0, \"AlterClientQuotasRequest\": 0, \"Heartbeat\": 0}, zbuf_grow: 0, buf_grow: 0, wakeups: Some(0), connects: Some(0), disconnects: Some(0), int_latency: Some(Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 11376, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }), outbuf_latency: Some(Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 11376, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }), rtt: Some(Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 13424, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }), throttle: Some(Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 17520, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }), toppars: {} }, \"iox-kafka-2.iox-kafka-brokers.iox.svc:9092/2\": Broker { name: \"iox-kafka-2.iox-kafka-brokers.iox.svc:9092/2\", nodeid: 2, nodename: \"iox-kafka-2.iox-kafka-brokers.iox.svc:9092\", source: \"learned\", state: \"UP\", stateage: 418926885924, outbuf_cnt: 0, outbuf_msg_cnt: 0, waitresp_cnt: 1, waitresp_msg_cnt: 0, tx: 839341, txbytes: 103067572, txerrs: 0, txretries: 0, req_timeouts: 0, rx: 839340, rxbytes: 86843970, rxerrs: 0, rxcorriderrs: 0, rxpartial: 0, req: {\"DescribeUserScramCredentialsRequest\": 0, \"DescribeClientQuotasRequest\": 0, \"Metadata\": 2095, \"ApiVersion\": 1, \"JoinGroup\": 0, \"OffsetFetch\": 0, \"AlterClientQuotasRequest\": 0, \"Heartbeat\": 0, \"FindCoordinator\": 701, \"Fetch\": 836543, \"Offset\": 1, \"SyncGroup\": 0, \"OffsetDeleteRequest\": 0, \"SaslHandshake\": 0, \"LeaveGroup\": 0, \"OffsetCommit\": 0, \"SaslAuthenticate\": 0}, zbuf_grow: 0, buf_grow: 0, wakeups: Some(1680142), connects: Some(1), disconnects: Some(0), int_latency: Some(Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 11376, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }), outbuf_latency: Some(Window { min: 14, max: 19, avg: 16, sum: 500, cnt: 30, stddev: 1, hdrsize: 11376, p50: 17, p75: 17, p90: 18, p95: 19, p99: 19, p99_99: 19, outofrange: 0 }), rtt: Some(Window { min: 499530, max: 501037, avg: 500292, sum: 15008779, cnt: 30, stddev: 510, hdrsize: 17520, p50: 501759, p75: 501759, p90: 501759, p95: 501759, p99: 501759, p99_99: 501759, outofrange: 0 }), throttle: Some(Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 17520, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }), toppars: {\"2bc31ebe26081a78_e124a0e40c5212c2-0\": TopicPartition { topic: \"2bc31ebe26081a78_e124a0e40c5212c2\", partition: 0 }} }, \"GroupCoordinator\": Broker { name: \"GroupCoordinator\", nodeid: 2, nodename: \"iox-kafka-2.iox-kafka-brokers.iox.svc:9092\", source: \"logical\", state: \"UP\", stateage: 116406978, outbuf_cnt: 0, outbuf_msg_cnt: 0, waitresp_cnt: 0, waitresp_msg_cnt: 0, tx: 699, txbytes: 27960, txerrs: 0, txretries: 0, req_timeouts: 0, rx: 699, rxbytes: 292182, rxerrs: 698, rxcorriderrs: 0, rxpartial: 0, req: {\"ApiVersion\": 699, \"Offset\": 0, \"SaslAuthenticate\": 0, \"DescribeClientQuotasRequest\": 0, \"DescribeUserScramCredentialsRequest\": 0, \"SaslHandshake\": 0, \"AlterClientQuotasRequest\": 0, \"FindCoordinator\": 0, \"Metadata\": 0, \"Heartbeat\": 0, \"OffsetDeleteRequest\": 0, \"LeaveGroup\": 0, \"JoinGroup\": 0, \"SyncGroup\": 0, \"Fetch\": 0, \"OffsetFetch\": 0, \"OffsetCommit\": 0}, zbuf_grow: 0, buf_grow: 0, wakeups: Some(2796), connects: Some(699), disconnects: Some(698), int_latency: Some(Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 11376, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }), outbuf_latency: Some(Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 11376, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }), rtt: Some(Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 13424, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }), throttle: Some(Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 17520, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }), toppars: {} }, \"iox-kafka-bootstrap:9092/bootstrap\": Broker { name: \"iox-kafka-bootstrap:9092/bootstrap\", nodeid: -1, nodename: \"iox-kafka-bootstrap:9092\", source: \"configured\", state: \"INIT\", stateage: 418327049180, outbuf_cnt: 0, outbuf_msg_cnt: 0, waitresp_cnt: 0, waitresp_msg_cnt: 0, tx: 4, txbytes: 229, txerrs: 0, txretries: 0, req_timeouts: 0, rx: 4, rxbytes: 854, rxerrs: 1, rxcorriderrs: 0, rxpartial: 0, req: {\"AlterClientQuotasRequest\": 0, \"DescribeUserScramCredentialsRequest\": 0, \"ApiVersion\": 1, \"Heartbeat\": 0, \"Fetch\": 0, \"FindCoordinator\": 2, \"LeaveGroup\": 0, \"JoinGroup\": 0, \"Offset\": 0, \"SyncGroup\": 0, \"OffsetDeleteRequest\": 0, \"DescribeClientQuotasRequest\": 0, \"OffsetFetch\": 0, \"Metadata\": 1, \"OffsetCommit\": 0, \"SaslAuthenticate\": 0, \"SaslHandshake\": 0}, zbuf_grow: 0, buf_grow: 0, wakeups: Some(11), connects: Some(1), disconnects: Some(1), int_latency: Some(Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 11376, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }), outbuf_latency: Some(Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 11376, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }), rtt: Some(Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 13424, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }), throttle: Some(Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 17520, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }), toppars: {} }}, topics: {\"2bc31ebe26081a78_e124a0e40c5212c2\": Topic { topic: \"2bc31ebe26081a78_e124a0e40c5212c2\", metadata_age: 116402, batchsize: Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 14448, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }, batchcnt: Window { min: 0, max: 0, avg: 0, sum: 0, cnt: 0, stddev: 0, hdrsize: 8304, p50: 0, p75: 0, p90: 0, p95: 0, p99: 0, p99_99: 0, outofrange: 0 }, partitions: {0: Partition { partition: 0, broker: 2, leader: 2, desired: true, unknown: false, msgq_cnt: 0, msgq_bytes: 0, xmit_msgq_cnt: 0, xmit_msgq_bytes: 0, fetchq_cnt: 0, fetchq_size: 0, fetch_state: \"active\", query_offset: -2, next_offset: 0, app_offset: -1001, stored_offset: -1001, committed_offset: -1001, eof_offset: 0, lo_offset: 0, hi_offset: 0, ls_offset: 0, consumer_lag: -1, txmsgs: 0, txbytes: 0, rxmsgs: 0, rxbytes: 0, msgs: 0, rx_ver_drops: 0, msgs_inflight: 0, next_ack_seq: 0, next_err_seq: 0, acked_msgid: 0 }, -1: Partition { partition: -1, broker: -1, leader: -1, desired: false, unknown: false, msgq_cnt: 0, msgq_bytes: 0, xmit_msgq_cnt: 0, xmit_msgq_bytes: 0, fetchq_cnt: 0, fetchq_size: 0, fetch_state: \"none\", query_offset: -1001, next_offset: 0, app_offset: -1001, stored_offset: -1001, committed_offset: -1001, eof_offset: -1001, lo_offset: -1001, hi_offset: -1001, ls_offset: -1001, consumer_lag: -1, txmsgs: 0, txbytes: 0, rxmsgs: 0, rxbytes: 0, msgs: 0, rx_ver_drops: 0, msgs_inflight: 0, next_ack_seq: 0, next_err_seq: 0, acked_msgid: 0 }} }}, cgrp: Some(ConsumerGroup { state: \"up\", stateage: 116402, join_state: \"init\", rebalance_age: 0, rebalance_cnt: 0, rebalance_reason: \"\", assignment_size: 0 }), eos: None }" log.target=rdkafka::client log.module_path=rdkafka::client log.file=/usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/client.rs log.line=85 target="log" time=1636805420272285242
```

# Proposed change
1. Make such message "debug" level
2. If there is a useful signal in these statistics, perhaps we can log the specific information that is useful

### PR DESCRIPTION
**Pull request by crepererum**
This maps a subset of Kafka stats as metrics. The set can -- of course
-- be changed in the future depending on our needs.

Fixes #3100.

### CODE REVIEW

File: write_buffer/src/config.rs [hunk 1]
```diff
@@ -456,13 +455,18 @@ mod tests {
...
+        let factory = factory();
 
         let state =
             MockBufferSharedState::empty_with_n_sequencers(NonZeroU32::try_from(1).unwrap());
         let mock_name = "some_mock";
         factory.register_always_fail_mock(mock_name.to_string());
         factory.register_mock(mock_name.to_string(), state);
     }
+
+    fn factory() -> WriteBufferConfigFactory {
```
Comment (alamb) [inline write_buffer/src/config.rs:467]: 👍

File: write_buffer/src/kafka.rs [hunk 2]
```diff
@@ -488,6 +495,88 @@ async fn maybe_auto_create_topics(
...
+            consumer_lag: metric_registry.register_metric(
+                "kafka_consumer_lag",
+                "The difference between `hi_offset` and `max(app_offset,committed_offset)`.",
+            ),
+        }
+    }
+}
+
+impl ClientContext for ClientContextImpl {
+    fn stats(&self, statistics: rdkafka::Statistics) {
```
Comment (alamb) [inline write_buffer/src/kafka.rs:537]: I double checked and indeed the default implementation of `ClientContext::stats` is what appears to be logging the message in #3100  -- aka https://docs.rs/rdkafka/0.27.0/src/rdkafka/client.rs.html#84-86 👍 

I also wondered if we needed to if we needed to implement the `log` here? The docs https://docs.rs/rdkafka/0.27.0/rdkafka/client/trait.ClientContext.html#method.log
 say 

> The default implementation forwards the log lines to the appropriate log crate macro. Consult the RDKafkaLogLevel documentation for details about the log level mapping.

However, when I looked at [the logs](https://logs.aws.influxdata.io/search?q=kubernetes_container_name%3Aiox+AND+librdkafka&rangetype=relative&relative=28800)  there were only a few messages, and they all looked like important things to be logging 👍

File: write_buffer/src/kafka.rs [hunk 3]
```diff
@@ -488,6 +495,88 @@ async fn maybe_auto_create_topics(
...
+        let attributes = metric::Attributes::from([
+            ("database", self.database_name.clone().into()),
+            ("client_type", statistics.client_type.into()),
+        ]);
+
+        self.producer_queue_msg_count
```
Comment (alamb) [inline write_buffer/src/kafka.rs:543]: I looked at the list of statistics in https://docs.rs/rdkafka/0.27.0/rdkafka/statistics/struct.Statistics.html and this seems to be a pretty good list 👍 

Maybe `msg_size` and `max_msg_size` might be worth adding as well

File: write_buffer/src/kafka.rs [hunk 4]
```diff
@@ -783,4 +880,46 @@ mod tests {
...
+
+    #[tokio::test]
+    async fn metrics() {
+        let conn = maybe_skip_kafka_integration!();
+        let adapter = KafkaTestAdapter::new(conn);
+        let ctx = adapter.new_context(NonZeroU32::new(1).unwrap()).await;
+
+        let reader = ctx.reading(true).await.unwrap();
+
+        // It seems that th reader must be used for rdkafka / rdkafka-rs to do anything at all =/
```
Comment (alamb) [inline write_buffer/src/kafka.rs:892]:
```suggestion
        // It seems that the reader must be used for rdkafka / rdkafka-rs to do anything at all =/
```

**Review by alamb — APPROVED** [review]
Looks good to me. This is very cool @crepererum

**Comment by crepererum** [comment]
The test currently deadlocks during process exit (but only if you run multiple tests), need to figure out what the hell is broken here.

**Comment by crepererum** [comment]
Fixed the "deadlock", see `fix: do not overload rdkafka w/ statistics`.

File: write_buffer/src/kafka.rs (hunk 2 above)
Comment (crepererum) [reply write_buffer/src/kafka.rs:537]: Yeah, I would keep the logging for now but we have the option to use our own logging if we want (e.g. to generate better structured logs).

File: write_buffer/src/kafka.rs (hunk 3 above)
Comment (crepererum) [reply write_buffer/src/kafka.rs:543]: `msg_size` is already here, I just renamed it to `producer_queue_msg_bytes` to better reflect what it actually contains.

`max_msg_size` doesn't exist. I THINK you want me to add `msg_max` and `msg_size_max` (aka the thresholds for `producer_queue_msg_count` and `producer_queue_msg_bytes`), right?

File: write_buffer/src/kafka.rs [hunk 5]
```diff
@@ -738,7 +738,12 @@ mod tests {
         }
 
         fn connection_config(&self) -> BTreeMap<String, String> {
-            BTreeMap::from([("statistics.interval.ms".to_owned(), "100".to_owned())])
+            BTreeMap::from([
```
Comment (alamb) [inline write_buffer/src/kafka.rs:741]: I don't understand this comment / fix -- is the thread really overloaded sending statistics 10x a second?

Getting statistics once a second seems fine to me too, something just seems strange 🤷

**Review by alamb — APPROVED** [review]

File: write_buffer/src/kafka.rs (hunk 5 above)
Comment (crepererum) [reply write_buffer/src/kafka.rs:741]: That's correct. I don't know what overloads it so heavily, but it seems to also occur when we just use the logging-driven stats impl (the code prior to this PR). Maybe stats are just way more work for rdkafka than we thought :shrug:

Comment (alamb) [reply write_buffer/src/kafka.rs:741]: I think this one is ready to go then

File: write_buffer/src/kafka.rs (hunk 3 above)
Comment (crepererum) [reply write_buffer/src/kafka.rs:556]: done

**Pull request merged**